File: python/station-distance-py/geo_services.py
from math import sin, cos, sqrt, atan2, radians


class GeoServices:
    """
    Collection of geo services
    """

    @staticmethod
    def get_distance_of_coords(point_a: tuple[float, float], point_b: tuple[float, float], unit: str = 'km') -> float:
        """
        Returns distance between two coordinates.
        The distance is calculated by the haversine formular

        Parameters
        ----------
        point_a : tuple[float, float]
            Coordinates (longitude, latitude) of point a
        point_b : tuple[float, float]
            Coordinates (longitude, latitude) of point b
        unit : str: optional
            Identifier for distance unit: 'km', 'mile'.
            Defaults to 'km'

        Returns
        -------
        float
            Distance between the two point in the given unit
        """

        # approximate radius of earth
        _R: float = 6373.0

        # Extract coordinate values as radians
        lon_a: radians = radians(point_a[0])
        lat_a: radians = radians(point_a[1])
        lon_b: radians = radians(point_b[0])
        lat_b: radians = radians(point_b[1])

        # Haversine formular
        d_lon: float = lon_b - lon_a
        d_lat: float = lat_b - lat_a

        a: float = sin(d_lat / 2)**2 + cos(lat_a) * cos(lat_b) * sin(d_lon / 2)**2
        d: float = 2 * atan2(sqrt(a), sqrt(1 - a))

        # distance in km
        distance: float = _R * d

        match unit:
            case 'km':
                return distance
            case 'mile':
                return distance/1.609344
            case default:
                return distance

File: kotlin/station-distance/src/main/kotlin/stationdistance/GeoServices.kt
package stationdistance

import kotlin.math.*

/**
 * Collection of geo services
 */
class GeoServices {

    companion object {
        // approximate radius of earth
        private const val R: Double = 6373.0;

        /**
         * Returns distance between two coordinates.
         * The distance is calculated by the haversine formular
         *
         * @param pointA Coordinates (longitude, latitude) of point a
         * @param pointB Coordinates (longitude, latitude) of point b
         * @param unit Identifier for distance unit: 'km', 'mile'. Defaults to 'km'.
         * @return Distance between the two point in the given unit rounded to Int.
         */
        fun getDistanceOfCoords(pointA: Pair<Double, Double>,
                                pointB: Pair<Double, Double>,
                                unit: String = "km") : Int {

            // Extract coordinate values as radians
            val lonA: Radian = pointA.first.toRadian();
            val latA: Radian = pointA.second.toRadian();
            val lonB: Radian = pointB.first.toRadian();
            val latB: Radian = pointB.second.toRadian();

            // Haversine formular
            val dLon: Double = lonB - lonA;
            val dLat: Double = latB - latA;

            val a: Double = sin(dLat / 2).pow(2) + cos(latA) * cos(latB) * sin(dLon/2).pow(2);
            val d: Double = 2 * atan2(sqrt(a), sqrt(1 - a));

            // distance in km
            val distance: Double = R * d;

            return when (unit) {
                "km" -> round(distance).toInt()
                "mile" -> round((distance/1.609344)).toInt()
                else -> round(distance).toInt()
            }
        }
    }
}

// Radian alias with function to convert degree to radian
typealias Radian = Double
fun Double.toRadian(): Radian = this / 180 * Math.PI